File: 1605-minimum-number-of-days-to-make-m-bouquets/minimum-number-of-days-to-make-m-bouquets.java
class Solution {
    public int minDays(int[] bloom, int m, int k) {
        
            if ((long)(m) * k > bloom.length) return -1;

        int l = Integer.MAX_VALUE, u = Integer.MIN_VALUE;
        for (int day : bloom) {
            l = Math.min(l, day);
            u = Math.max(u, day);
        }

        int ans = -1;
        while (l <= u) {
            int mid = l + (u - l) / 2;
            if (canMake(bloom, mid, m, k)) {
                ans = mid;
                u = mid - 1;
            } else {
                l = mid + 1;
            }
        }
        return ans;
    }

    private boolean canMake(int[] b, int day, int m, int k) {
        int bouquets = 0, flowers = 0;
        for (int value : b) {
            if (value <= day) {
                flowers++;
                if (flowers == k) {
                    bouquets++;
                    flowers = 0;
                }
            } else {
                flowers = 0;
            }
        }
        return bouquets >= m;
    }

}

File: 1605-minimum-number-of-days-to-make-m-bouquets/minimum-number-of-days-to-make-m-bouquets.cpp
class Solution {
public:
    int minDays(vector<int>& bloom, int m, int k) {
         int l=*min_element(bloom.begin(),bloom.end()),u=*max_element(bloom.begin(),bloom.end()),n=bloom.size();
         if((long long)m*k>n)
         return -1;
       return calc_days(bloom,m,k,l,u);
    }
    int calc_days(vector<int>&b,int m,int k,int l,int u)
    {
        while(l<=u)
        {
            int mid=l+(u-l)/2;
            if(can_make(b,mid,k)>=m)
            {   
                u=mid-1;
            }
            else
            l=mid+1;
        }
        return l;
    }
    int can_make(vector<int>&b,int day,int k)
    { long long tm=0;
      long long c=0;
      for(int i=0;i<b.size();i++)
      {
        if(b[i]<=day)
        {
            c++;
        }
        else
        {
            tm+=c/k;
            c=0;
        }
         
      }
      tm+=c/k;
      return tm;
    }    
    
};